### ISSUE TITLE
`system.etc.overlay` doesn't create /etc/NIXOS file

### ISSUE THREAD

**Issue by name-snrl** [NixOS/nixpkgs#341453]
### Describe the bug
Today I was curious to test the `system.etc.overlay.enable option. After enabling it and rebooting, I could not find the /etc/NIXOS file, which to my knowledge is used to verify that the operating system is NIXOS.

### Steps To Reproduce
1. ...
2. ...
3. ...

### Notify maintainers

@nikstur @Mic92 @roberth 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
fish > nix-shell -p nix-info --run "nix-info -m"
fetching path input 'path:/nix/store/5w3dp0m37794iffsbm9vd9f1xmmhda6i-source'
 - system: `"x86_64-linux"`
 - host os: `Linux 6.10.7-zen1, NixOS, 24.11 (Vicuna), 24.11.20240906.574d1ea`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.0
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/name_snrl/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/name_snrl/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/name_snrl/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/name_snrl/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/dzs6wmlzg8hdh9mjailpghvxrzs43jdn-lix-2.91.0/share`
 - nixpkgs: `/nix/store/5w3dp0m37794iffsbm9vd9f1xmmhda6i-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by jopejoe1** [comment]
I found 22 references to `/etc/NIXOS` in nixpkgs which can be filtered down to steam, zsh, switch-to-configuration and various places that create it, not sure if any out of tree uses exit.

<details>
<summary>22 References of `/etc/NIXOS`</summary>
<pre><code>
pkgs/shells/zsh/default.nix:if test -e /etc/NIXOS; then
pkgs/games/steam/fhsenv.nix:    if [ -f /etc/NIXOS ]; then   # Check only useful on NixOS
pkgs/games/steam/fhsenv.nix:    "--bind /etc/NIXOS /etc/NIXOS" # required 32bit driver check in runScript
configuration-ng/src/src/main.rs:    // This is a NixOS installation if it has /etc/NIXOS or a proper /etc/os-release.
pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:    if !Path::new("/etc/NIXOS").is_file()
nixos/tests/switch-test.nix:    # when /etc/NIXOS is missing. If the parser does not work, switch-to-configuration
nixos/tests/switch-test.nix:    machine.succeed("rm /etc/NIXOS")
nixos/modules/system/boot/stage-1-init.sh:    touch "$root/etc/NIXOS"
nixos/modules/system/etc/setup-etc.pl:# Create /etc/NIXOS tag if not exists.
nixos/modules/system/etc/setup-etc.pl:open TAG, ">>/etc/NIXOS";
nixos/modules/system/activation/switch-to-configuration.pl:# This is a NixOS installation if it has /etc/NIXOS or a proper
nixos/modules/system/activation/switch-to-configuration.pl:if (!-f "/etc/NIXOS" && (read_file("/etc/os-release", err_mode => "quiet") // "") !~ /^ID="?@distroId@"?/msx) {
nixos/modules/installer/cd-dvd/iso-image.nix:        # /etc/NIXOS tag.
nixos/modules/installer/cd-dvd/iso-image.nix:        touch /etc/NIXOS
nixos/modules/installer/netboot/netboot.nix:        # /etc/NIXOS tag.
nixos/modules/installer/netboot/netboot.nix:        touch /etc/NIXOS
nixos/modules/installer/sd-card/sd-image.nix:        # nixos-rebuild also requires a "system" profile and an /etc/NIXOS tag.
nixos/modules/installer/sd-card/sd-image.nix:        touch /etc/NIXOS
nixos/doc/manual/installation/installing-from-other-distro.section.md:1.  Set up the `/etc/NIXOS` and `/etc/NIXOS_LUSTRATE` files:
nixos/doc/manual/installation/installing-from-other-distro.section.md:    `/etc/NIXOS` officializes that this is now a NixOS partition (the
nixos/doc/manual/installation/installing-from-other-distro.section.md:    $ sudo touch /etc/NIXOS
nixos/doc/manual/installation/installing-from-other-distro.section.md:    $ echo etc/nixos | sudo tee -a /etc/NIXOS_LUSTRATE
</code></pre>
</details>

**Comment by name-snrl** [comment]
> not sure if any out of tree uses exit

yes, and usually if /etc/NIXOS doesn't exist, it fallbacks to /etc/os-release, as shown in this test case:

https://github.com/NixOS/nixpkgs/blob/e1b4ce3b8166285fb1f982886e48ca0c74d4bd8f/nixos/tests/switch-test.nix#L658-L660

anyway, it's probably used outside of nixpkgs, in some custom scripts. shouldn't we keep compatibility?

**Comment by Mic92** [comment]
Could we not create /etc/NIXOS in a declarative way instead of setup-etc.pl?

**Comment by Mic92** [comment]
But it also seems we are not actually relying anymore on `/etc/NIXOS` in-tree.

**Comment by Mic92** [comment]
steam-run still needs it. And nixos-install creates only /etc/NIXOS but not /etc/os-release.

**Comment by nikstur** [comment]
We could probably just set it via `environment.etc`, right? I don't see a reason why we couldn't.

**Comment by arianvp** [comment]
`nixos-enter` still relies on `/etc/NIXOS` (But could be fixed).

**Comment by nyabinary** [comment]
> `nixos-enter` still relies on `/etc/NIXOS` (But could be fixed).

How?

**Comment by 1sixth** [comment]
FYI, VS Code [checks](https://github.com/microsoft/vscode/blob/d9637b3f2faa8f8ce0636e556291fdb1ba714c0b/cli/src/util/prereqs.rs#L175) whether `/etc/NIXOS` exists to avoid mandating glibc versions. I encountered this issue while trying to use VS Code Remote-SSH after enabling `system.etc.overlay`.

Should we set `/etc/NIXOS` by default?
